lightwood/encoders/text/pretrained.py

```diff
@@ -0,0 +1,323 @@
+        device - torch.device; cuda/cpu
+        log - lightwood.logger.log; print output
+        optim - transformers.optimization.AdamW; optimizer
+        n_epochs - number of epochs to train
+
+        """
+        self._model.train()
+
+        if optim is None:
+            print("No opt. provided, setting all params with AdamW.", flush=True)
+            optim = AdamW(self._model.parameters(), lr=5e-5)
+        else:
+            print("Optimizer provided", flush=True)
+
+        if scheduler is None:
+            print("No scheduler provided.", flush=True)
+        else:
+            print("Scheduler provided.", flush=True)
+
+        print("Beginning training.")
+        for epoch in range(n_epochs):
+            total_loss = 0
+
+            for batch in dataset:
+                optim.zero_grad()
+
+                with LightwoodAutocast():
+                    inpids = batch["input_ids"].to(self.device)
+                    attn = batch["attention_mask"].to(self.device)
+                    labels = batch["labels"].to(self.device)
+                    outputs = self._model(inpids, attention_mask=attn, labels=labels)
+                    loss = outputs[0]
+
+                total_loss += loss.item()
+
+                loss.backward()
+                optim.step()
+                if scheduler is not None:
+                    scheduler.step()
+
+            # self._train_callback(epoch, loss.item())
+            print("Epoch=", epoch + 1, "Loss=", total_loss / len(dataset), flush=True)
+
+    def _train_callback(self, epoch, loss):
+        log.info(f"{self.name} at epoch {epoch+1} and loss {loss}!")
+
+    def encode(self, column_data):
+        """
+        TODO: Maybe batch the text up; may take too long
+        Given column data, encode the dataset.
+
+        Currently, returns the embedding of the pre-classifier layer.
+
+        Args:
+        column_data:: [list[str]] list of text data in str form
+
+        Returns:
+        encoded_representation:: [torch.Tensor] N_sentences x Nembed_dim
+        """
+        if self._prepared is False:
+            raise Exception("You need to first prepare the encoder.")
+
+        # Freeze training mode while encoding
+        self._model.eval()
+
+        encoded_representation = []
+
+        with torch.no_grad():
+            # Set the weights; this is GPT-2
+            for text in column_data:
+
+                # Omit NaNs
+                if text == None:
+                    text = ""
+
+                # Tokenize the text with the built-in tokenizer.
+                inp = self._tokenizer.encode(
+                    text, truncation=True, return_tensors="pt"
+                ).to(self.device)
+
+                output = self._model.base_model(inp).last_hidden_state[:, 0]
+
+                # If the model has a pre-classifier layer, use this embedding.
+                if hasattr(self._model, "pre_classifier"):
+                    output = self._model.pre_classifier(output)
```

Comment:
Wait... wasn't the idea to concat the classifier layer output with the embedding? This seems to be replacing it.

Reply:
The benchmarks I've done used explicitly this layer's output; the pre-classifier layer is just a linear layer, hence the output is actually Nbatch x Nembedding in dim (it's a 1:1 from the transformer output embedding size but only uses the CLS token). You can think of the pre-classifier layer like a "head" that modifies the output of the transformer. In theory, one linear layer that directly maps into the output probably is ok, but the original idea is that this intermediate layer could be tweaked into a lower dimensional representation instead of being 1:1. 

If we concat the output of the embedding to the transformer, we would get a vector thats 768*2 (768 - embedding dim); given the tuning approach changes the output of transformer params, I'm not sure why the embedding itself is needed. Maybe you're thinking of the prior approach of which the output was just the target prediction?